costmap_cspace/include/costmap_cspace/costmap_3d_layer/footprint.h

```diff
@@ -228,31 +235,54 @@ class Costmap3dLayerFootprint : public Costmap3dLayerBase
     // Get max
     for (size_t yaw = 0; yaw < map->info.angle; yaw++)
     {
-      for (unsigned int i = 0; i < msg->data.size(); i++)
+      for (size_t i = 0; i < msg->data.size(); i++)
       {
         const int gx = lroundf((i % msg->info.width) * msg->info.resolution / map->info.linear_resolution) + ox;
         const int gy = lroundf((i / msg->info.width) * msg->info.resolution / map->info.linear_resolution) + oy;
-        if ((unsigned int)gx >= map->info.width ||
-            (unsigned int)gy >= map->info.height)
+        if (static_cast<size_t>(gx) >= map->info.width ||
+            static_cast<size_t>(gy) >= map->info.height)
           continue;
-
-        auto val = msg->data[i];
-        if (val <= 0)
+        const int8_t val = msg->data[i];
+        if (val < 0)
+        {
           continue;
+        }
+        else if (val == 0)
+        {
+          int8_t& m = map->getCost(gx, gy, yaw);
+          if (m < 0)
+            m = 0;
+          continue;
+        }
 
         for (int y = -range_max_; y <= range_max_; y++)
         {
           for (int x = -range_max_; x <= range_max_; x++)
           {
             const int x2 = gx + x;
             const int y2 = gy + y;
-            if (static_cast<unsigned int>(x2) >= map->info.width ||
-                static_cast<unsigned int>(y2) >= map->info.height)
+            if (static_cast<size_t>(x2) >= map->info.width ||
+                static_cast<size_t>(y2) >= map->info.height)
               continue;
 
-            auto& m = map->getCost(x2, y2, yaw);
-            const auto c = cs_template_.e(x, y, yaw) * val / 100;
-            if (m < c)
+            if (keep_unknown_)
+            {
+              const int mx = lroundf((x2 - ox) * map->info.linear_resolution / msg->info.resolution);
+              const int my = lroundf((y2 - oy) * map->info.linear_resolution / msg->info.resolution);
+
+              // Out of the updated map is unknown
+              if (static_cast<size_t>(mx) >= msg->info.width ||
+                  static_cast<size_t>(my) >= msg->info.height)
+                continue;
+
+              const size_t i2 = my * msg->info.width + mx;
+              if (msg->data[i2] < 0)
```

Comment:
The behaviour has been fixed, but computation cost increases too much by this. (25% -> 80%)